src/fs/metadata_log.hh

```diff
@@ -0,0 +1,275 @@
+    // E.g.
+    // To create file, _create_or_delete_lock is acquired first, then corresponding dir entry is unique-locked and finally
+    // file creation is performed (notice that we do not have to acquire shared lock on the directory to which we add entry because
+    // we hold global lock, so it won't disappear in the meantime).
+    // To make a read or write to a file, a shared lock is acquired on its inode and then the operation is performed.
+    // Motivation:
+    // Global unique lock ensures that only one operation using unique locks on inode / dir entry takes place at any time.
+    // "Local" unique locks ensure that resource is not used by anyone else. Regarding deadlocks, the only place for
+    // a deadlock to appear is while unique lock is taken on an inode / dir entry. As far as, the operation holding global lock
+    // acquires only one locks it's all good. If two are acquired at the same time, an extreme caution should be used and
+    // if possible two unique locks should be avoided.
+    value_shared_lock<inode_t> _inode_locks;
+    value_shared_lock<std::pair<inode_t, sstring>> _dir_entry_locks;
+    semaphore _create_or_delete_lock {1};
+
+    // TODO: for compaction: keep some set(?) of inode_data_vec, so that we can keep track of clusters that have lowest utilization (up-to-date data)
+    // TODO: for compaction: keep estimated metadata log size (that would take when written to disk) and the real size of metadata log taken on disk to allow for detecting when compaction
+
+    friend class metadata_log_bootstrap;
+    friend class create_file_operation;
+
+public:
+    metadata_log(block_device device, unit_size_t cluster_size, unit_size_t alignment);
+
+    metadata_log(const metadata_log&) = delete;
+    metadata_log& operator=(const metadata_log&) = delete;
+
+    future<> bootstrap(inode_t root_dir, cluster_id_t first_metadata_cluster_id, cluster_range available_clusters, fs_shard_id_t fs_shards_pool_size, fs_shard_id_t fs_shard_id);
+
+    void write_update(inode_info::file& file, inode_data_vec new_data_vec);
+
+    // Deletes data vectors that are subset of @p data_range and cuts overlapping data vectors to make them not overlap
+    void cut_out_data_range(inode_info::file& file, file_range range);
+
+private:
+    void memory_only_create_inode(inode_t inode, bool is_directory, unix_metadata metadata);
+
+    void memory_only_update_metadata(inode_t inode, unix_metadata metadata);
+
+    void memory_only_delete_inode(inode_t inode);
+
+    void memory_only_small_write(inode_t inode, disk_offset_t offset, temporary_buffer<uint8_t> data);
+
+    void memory_only_update_mtime(inode_t inode, decltype(unix_metadata::mtime_ns) mtime_ns);
+
+    void memory_only_truncate(inode_t inode, disk_offset_t size);
+
+    void memory_only_add_dir_entry(inode_info::directory& dir, inode_t entry_inode, sstring entry_name);
+
+    void memory_only_delete_dir_entry(inode_info::directory& dir, sstring entry_name);
```

Comment:
nitpick: it's fine for the declarations above to be separated by a single newline only, not double.